### PR TITLE
[Bugfix] Track DFlash Mamba state at checkpoint boundaries

### LINKED ISSUE #37817: [Bug] DFlash misses Mamba checkpoints when accepted tokens cross a tracking boundary
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

We hit a cache consistency issue with Qwen3.6-35B-A3B using DFLASH and `mamba-radix-cache-strategy=extra_buffer`.

The model has both full-attention KV cache and linear-attention state. The linear state is checkpointed at fixed boundaries (256 tokens in our setup). If a DFlash verify step crosses one of these boundaries and then rolls back rejected tokens, the active linear state is restored to the last accepted token, but the state at the checkpoint boundary is not copied to the tracking buffer.

The problem appears to be in `DFlashWorkerV2._update_target_mamba_state_after_verify`:

```python
to_track_mask = (
    seq_lens_pre_verify // mamba_track_interval
    != batch.seq_lens // mamba_track_interval
)
```

`seq_lens_pre_verify` is cloned from `batch.seq_lens`, and `batch.seq_lens` is not advanced before this check. Only `batch.seq_lens_cpu` is temporarily changed for backend planning. As a result, the two sides above use the same length and no boundary is detected.

For example, with an interval of 256:

- length before verify: 255
- verify block size: 16
- committed tokens after verification: 1
- resulting committed length: 256

Step 0 should be saved as the state checkpoint at length 256, but `mamba_steps_to_track` is set to -1.

The scheduler-side result handling does detect the accepted range crossing the boundary and updates `mamba_last_track_seqlen` and the ping-pong buffer bookkeeping. This can leave a radix-cache entry whose full-attention KV corresponds to length 256 while its linear-attention state is stale.

The shared speculative path in `spec_utils._verify_commit_step_indices` calculates the post-verify length from the accepted length:

```python
seq_lens_post_verify = batch.seq_lens + accept_lens
```

The DFlash path likely needs the equivalent calculation based on `seq_lens_pre_verify + commit_lens`.

Related: #35821 and #37326 cover nearby accepted-state tracking problems, but do not appear to cover this DFlash-specific calculation.

We found this during a joint investigation. The fix is ready for review in #37818.

### Reproduction

The boundary condition can be reduced to:

```python
import torch

interval = 256
pre = torch.tensor([255])
batch_seq_lens = pre.clone()
commit_lens = torch.tensor([1])

current = pre // interval != batch_seq_lens // interval
expected = pre // interval != (pre + commit_lens) // interval

assert current.item() is False
assert expected.item() is True
```

Configuration under test:

```text
Model: Qwen/Qwen3.6-35B-A3B
Speculative algorithm: DFLASH
Mamba radix cache strategy: extra_buffer
Mamba track interval: 256
```

Expected behavior for representative boundary cases:

- pre=255, commit=1: save step 0 at 256
- pre=254, commit=2: save step 1 at 256
- pre=250, commit=3: do not save a checkpoint
- pre=256, commit=1: do not save 256 again

### Environment

Repository: sgl-project/sglang
Branch: main
Commit: f5bed255c0d855a1ab98bd3a7df7297a33dca5e3
Model: Qwen/Qwen3.6-35B-A3B

Validation is complete. See #37818 for the fix, regression test, and GPU validation results.

### PR DESCRIPTION
**Pull request by simple-sun**
## Motivation

Fixes #37817.

DFlash can miss a linear-attention state checkpoint when accepted verify tokens cross a Mamba tracking boundary. At this point, `batch.seq_lens` still contains the pre-verify lengths, so the checkpoint-crossing check compares the pre-verify length against itself.

The scheduler continues to advance the sequence length, which can leave the full-attention KV cache and the linear-attention state at different positions.

## Modifications

- Compute the committed post-verify sequence lengths before committing the Mamba state.
- Use the post-verify lengths to detect checkpoint crossings and select the verify step whose state should be tracked.
- Add a lightweight regression test for the checkpoint bookkeeping logic, covering:
  - int32 and int64 sequence lengths;
  - tracking intervals 64, 128, and 256;
  - requests that cross a checkpoint boundary;
  - requests that do not cross a boundary.

The change is implemented in `DFlashWorkerV2`, so it applies to both DFlash and DFlash2 paths that use this worker.

### CPU test — checkpoint bookkeeping

Scope: check checkpoint-crossing detection, the verify step selected for tracking, and the last accepted step. The test uses CPU tensors and mocks the Mamba state update, so it does not load a model or execute CUDA kernels. It is registered for CPU CI.

Result: the unit test passes all six combinations of int32/int64 lengths and tracking intervals 64/128/256. Each combination includes crossed and uncrossed requests in the same batch, including the 255 → 256 crossing at interval 256.

### GPU tests — model behavior and request handling

Setup: Qwen3.6-35B-A3B and its DFlash draft model on 2×H20, TP=2, with Mamba extra-buffer cache.

Scope: run real model inference and compare the next generated token after a warm-cache prefix lookup against cold-cache inference of the same input. The same 56 frozen cases are used for unpatched and patched code, covering positions before, at, and after a checkpoint boundary, prefill-only prefixes, and concurrent boundary +1 requests. Boundary +1 means a prefix ending one token past the checkpoint.

For the comparisons below, a stable case is defined as one whose cold-path top-1 versus top-2 log-prob margin is at least 0.5. The complete 56-case result is also reported, so low-margin cases are not excluded from the overall result.

Results at tracking interval 256:

| Comparison | Unpatched | Patched |
| --- | ---: | ---: |
| All frozen cases | 24/56 mismatches | 2/56 mismatches |
| Stable subset | 21/46 mismatches | 0/46 mismatches |
| Boundary +1 | 8/8 mismatches | 0/8 mismatches |
| Concurrent boundary +1 | 7/8 mismatches | 0/8 mismatches |

The two remaining patched mismatches are repeated token flips from the same boundary -1 case. Its cold-path top-1 versus top-2 log-prob margin is 0.125, below the predefined stable threshold of 0.5.

This case does not cross a checkpoint boundary and therefore does not exercise the checkpoint-crossing path changed by this PR. It remains included in the complete 56-case result.

Load-test scope: check request completion and non-empty responses at tracking intervals 128 and 256, with concurrency levels 1, 4, 8, and 16. Each run contains 128 requests, including 64 streaming requests.

Load-test result: 128/128 requests completed successfully in every run, for a total of 1,024/1,024 requests, including 512 streaming requests.

## Speed Tests and Profiling

Not applicable. The change only replaces the sequence-length tensor used by the existing checkpoint bookkeeping.

## Collaboration

This fix was jointly developed and validated by @lvweiv and @simple-sun, including the root-cause analysis, implementation, and boundary regression testing.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #34657542229](https://github.com/sgl-project/sglang/actions/runs/34657542229)
Latest PR Test (Extra): :x: [Run #34657542015](https://github.com/sgl-project/sglang/actions/runs/34657542015)
Latest PR Test (AMD ROCm 10): :x: [Run #34657542130](https://github.com/sgl-project/sglang/actions/runs/34657542130)

### CODE REVIEW

**Comment by kpham-sgl** [comment]
/rerun-test test/registered/e2e/models/test_nvidia_nemotron_3_nano.py::TestNvidiaNemotron35LightningNVFP4DFlash test/registered/attention/unittests/gdn/test_flashinfer.py test/registered/attention/unittests/gdn/test_triton.py test/registered/attention/unittests/kda/test_triton.py test/registered/attention/unittests/lightning/test_triton.py test/registered/unit/spec/test_dflash_extra_buffer_lazy.py test/registered/unit/spec/test_dflash_overlap_hostsync.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/e2e/models/test_nvidia_nemotron_3_nano.py::TestNvidiaNemotron35LightningNVFP4DFlash test/registered/attention/unittests/gdn/test_flashinfer.py test/registered/attention/unittests/gdn/test_triton.py test/registered/attention/unittests/kda/test_triton.py test/registered/attention/unittests/lightning/test_triton.py test/registered/unit/spec/test_dflash_extra_buffer_lazy.py test/registered/unit/spec/test_dflash_overlap_hostsync.py

**Review by kpham-sgl — APPROVED** [review]
